### ISSUE TITLE
UDFs for canonical naming of mobile applications

### ISSUE THREAD

**Issue by jklukas** [mozilla/bigquery-etl#947]
Mobile usage data that feeds into the mobile MAU KPI and into GUD relies on some ad hoc filtering in various views and scheduled queries to determine which pings contribute.

It's currently difficult to follow the trail of queries, and you need to copy and paste logic out of those queries to recreate this filtering and naming for new use cases.

I initially wanted to provision a table to join with to get this cleaned info, but I don't see a clear way to do that since canonical app names can depend on source ping type, channel, an `app_name` field, etc.

The interface would be something like the following:

```
udf.extract_canonical_name_core(app_name, os, channel)
RETURNS STRUCT<
  canonical_name STRING,
  canonical_channel STRING,
  contributes_to_2019_kpi BOOLEAN
  >

udf.extract_canonical_name_glean(dataset_name, channel)
RETURNS STRUCT<
  canonical_name STRING,
  canonical_channel STRING,
  contributes_to_2019_kpi BOOLEAN
  >
```

cc @jmccrosky

**Comment by jmccrosky** [comment]
I'm not totally clear on how this would be used.  Can you clarify?  @gkabbz was working on implementing the mobile activation metric that started this discussion, so may have some thoughts to share.

**Comment by jklukas** [comment]
Currently, we have logic in the `core_clients_last_seen` and `nondesktop_clients_last_seen` views that decides how to clean and filter app names.  We would refactor those so that they call these functions instead. Then the exact_mau and GUD queries would make use of those canonical names.

I may need to go ahead and implement this in a PR to make the discussion more concrete.

**Comment by jmccrosky** [comment]
I understand I think.  I guess generally I'm supportive of having a single clients_last_seen view that works across all our products and trust you to choose the best way to get there.  If you want me to have more opinions than just that though, let me know.

**Comment by jklukas** [comment]
I suppose my real intention here is to have a canonical place I can point to when the question comes up "which pings exactly are included for the mobile MAU KPI?" These UDFs would be the concrete implementation of that logic and would also serve as the most concrete documentation of the KPI.

**Comment by jmccrosky** [comment]
Sounds good to me :)

**Comment by magorlick** [comment]
This work is going to be critical moving forward.  The Mobile Data Org is regularly working through communications challenges and data integrity challenges trying to unite information across sources that take different approaches to cleaning channel, os and app_name.  Example: non_desktop_clients_last_seen vs mobile_search_clients_daily_v1.  The parameter names are different as well as the contents.  This creates uncertainty and mis-matched join keys can create biased insights that are not meaningful.  Most recently, confusion around what app_name connects to what product ID delayed the launch of an experiment and created several extra cycles to validate the data.  Ideally, a complete map or UDF is available so this is extremely easy to correctly label and trace back.

**Comment by fbertsch** [comment]
I believe `non_desktop_clients_last_seen` and `mobile_search_clients_daily` are both using the canonical naming: see [here for clients_last_seen](https://github.com/mozilla/bigquery-etl/blob/master/sql/telemetry/nondesktop_clients_last_seen_v1/view.sql#L9-L37) and [here for search_clients_daily](https://github.com/mozilla/bigquery-etl/blob/master/sql/search_derived/mobile_search_clients_daily_v1/mobile_search_clients_daily.sql.py#L8-L14).

@magorlick were those tables the cause of the delay?

**Comment by magorlick** [comment]
This specific join was not the problem, but the communication about which product is which across third party and internal tools, product_ids, table names, third party tokens, individuals etc. 

I looked at the links you sent me and, while the code seems ok, the property names for the join aren't always clear.  Example: in search_clients_daily app_name is 'clean' and normalized_app_name is not.  Using the word 'normalized' is interpreted to mean this is the correct clean join key adding confusing to analysis.

**Comment by Ben-Wu** [comment]
Firefox preview (org.mozilla.fenix) is now firefox nightly so we should start differentiating FF preview beta and FF nightly for historical data.  It looks like we can use `app_display_version` where nightly starts with `Nightly` and preview is a `x.y.z` version.

**Comment by acmiyaguchi** [comment]
This should be fixed by #1226.
